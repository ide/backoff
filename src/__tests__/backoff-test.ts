import { computeNextBackoffInterval } from '../Backoff';

it(`computes exponential backoff`, () => {
  expect(computeNextBackoffInterval(1, 0, { multiplier: 2, randomizationFactor: 0 })).toBe(1);
  expect(computeNextBackoffInterval(1, 1, { multiplier: 2, randomizationFactor: 0 })).toBe(2);
  expect(computeNextBackoffInterval(1, 2, { multiplier: 2, randomizationFactor: 0 })).toBe(4);
});

it(`applies random jitter`, () => {
  for (let i = 0; i < 10; i++) {
    const backoff = computeNextBackoffInterval(1, 2, { multiplier: 2, randomizationFactor: 0.5 });
    expect(backoff).toBeGreaterThanOrEqual(2);
    expect(backoff).toBeLessThanOrEqual(6);
  }
});

it(`supports constant backoff with a multiplier of one`, () => {
  expect(computeNextBackoffInterval(1, 0, { multiplier: 1, randomizationFactor: 0 })).toBe(1);
  expect(computeNextBackoffInterval(1, 1, { multiplier: 1, randomizationFactor: 0 })).toBe(1);
  expect(computeNextBackoffInterval(1, 2, { multiplier: 1, randomizationFactor: 0 })).toBe(1);
});

it(`honors a minimum backoff interval`, () => {
  const minBackoff = 2;
  expect(computeNextBackoffInterval(1, 0, { minBackoff })).toBe(minBackoff);
});

it(`honors a maximum backoff interval`, () => {
  const maxBackoff = 30;
  expect(computeNextBackoffInterval(1, 10, { maxBackoff })).toBe(maxBackoff);
});

it(`validates arguments`, () => {
  // Invalid default interval
  expect(() => computeNextBackoffInterval(-1, 0)).toThrowErrorMatchingSnapshot();

  // Invalid retry count
  expect(() => computeNextBackoffInterval(1, -1)).toThrowErrorMatchingSnapshot();

  // Invalid multiplier
  expect(() =>
    computeNextBackoffInterval(1, 0, { multiplier: 0.5 }),
  ).toThrowErrorMatchingSnapshot();

  // Invalid randomization factor
  expect(() =>
    computeNextBackoffInterval(1, 0, { randomizationFactor: -0.1 }),
  ).toThrowErrorMatchingSnapshot();
  expect(() =>
    computeNextBackoffInterval(1, 0, { randomizationFactor: 1.1 }),
  ).toThrowErrorMatchingSnapshot();

  // Invalid minimum interval
  expect(() => computeNextBackoffInterval(1, 0, { minBackoff: -1 })).toThrowErrorMatchingSnapshot();
});
