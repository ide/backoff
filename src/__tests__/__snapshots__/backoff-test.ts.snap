// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validates arguments 1`] = `"The initial backoff interval must be positive"`;

exports[`validates arguments 2`] = `"The previous retry count must not be negative"`;

exports[`validates arguments 3`] = `"The backoff multiplier must be greater than or equal to 1"`;

exports[`validates arguments 4`] = `"The randomization factor must be between 0 and 1, inclusive"`;

exports[`validates arguments 5`] = `"The randomization factor must be between 0 and 1, inclusive"`;

exports[`validates arguments 6`] = `"The minimum backoff interval must be positive"`;
